alf/data_structures.py

```diff
@@ -42,55 +40,83 @@ def namedtuple(typename, field_names, default_value=None, default_values=()):
+
+    A `TimeStep` contains the data emitted by an environment at each step of
+    interaction. A `TimeStep` holds a `step_type`, an `observation` (typically a
+    NumPy array or a dict or list of arrays), and an associated `reward` and
+    `discount`.
+
+    The first `TimeStep` in a sequence will equal `StepType.FIRST`. The final
+    `TimeStep` will equal `StepType.LAST`. All other `TimeStep`s in a sequence
+    will equal `StepType.MID.
+
+    Attributes:
+      step_type: a `Tensor` of `StepType` enum values.
+      reward: a `Tensor` of reward values from executing 'prev_action'.
+      discount: A discount value in the range `[0, 1]`.
+      observation: A (nested) 'Tensor' for observation
+      prev_action: A 'Tensor' for action from previous tiem step
+      env_id: the ID of the environment from which this time_step is
+    """
 
     def is_first(self):
-        if tf.is_tensor(self.step_type):
-            return tf.equal(self.step_type, StepType.FIRST)
-        return np.equal(self.step_type, StepType.FIRST)
+        if torch.is_tensor(self.step_type):
+            return torch.eq(self.step_type, StepType.FIRST)
+        raise ValueError('step_type is not a Torch Tensor')
 
     def is_mid(self):
-        if tf.is_tensor(self.step_type):
-            return tf.equal(self.step_type, StepType.MID)
-        return np.equal(self.step_type, StepType.MID)
+        if torch.is_tensor(self.step_type):
+            return torch.eq(self.step_type, StepType.MID)
+        raise ValueError('step_type is not a Torch Tensor')
 
     def is_last(self):
-        if tf.is_tensor(self.step_type):
-            return tf.equal(self.step_type, StepType.LAST)
-        return np.equal(self.step_type, StepType.LAST)
-
+        if torch.is_tensor(self.step_type):
+            return torch.eq(self.step_type, StepType.LAST)
+        raise ValueError('step_type is not a Torch Tensor')
 
-def make_action_time_step(time_step: TimeStep, prev_action, first_env_id=0):
-    """Make ActionTimeStep from TimeStep.
+    # def __hash__(self):
+    # TODO(b/130243327): Explore performance impact and consider converting
+    # dicts in the observation into ordered dicts in __new__ call.
+    # TODO(Jerry): wait for pytorch version of nest
+    # return hash(tuple(tf.nest.flatten(self)))
 
-    Args:
-        time_step (TimeStep):
-        prev_action (nested Tensor):
-        first_env_id (int): the environment ID for the first sample in this
-            batch. The environment IDs for the other samples are incremented
-            from this.
-    Returns:
-        ActionTimeStep
-    """
-    return ActionTimeStep(
-        step_type=time_step.step_type,
-        reward=time_step.reward,
-        discount=time_step.discount,
-        observation=time_step.observation,
-        prev_action=prev_action,
-        env_id=first_env_id + tf.range(time_step.step_type.shape[0]))
 
+PolicyStep = namedtuple('PolicyStep', ('action', 'state', 'info'))
```

Comment:
default_value=(). Otherwise the default_value is None, which cannot be handled by nest.